### PR TITLE
[sparse-fsm-encode] Improvements for Software

### PR DESCRIPTION
**Pull request by lenary**
This PR contains two changes:
- One to avoid using complementary encodings
- One to change how we represent these enums using Rust types.

<details><summary>Example <code>--language=rust</code> Output</summary>

```

------------------------------------------------
| COPY PASTE THE CODE TEMPLATE BELOW INTO YOUR |
| RUST FILE, INLUDING THE COMMENT IN ORDER TO  |
| EASE AUDITABILITY AND REPRODUCIBILITY.       |
------------------------------------------------

///```text
/// Encoding generated with
/// $ ./sparse-fsm-encode.py -d 10 -m 3 -n 16 \
///     -s 2855134562 --language=rust
///
/// Hamming distance histogram:
///
///  0: --
///  1: --
///  2: --
///  3: --
///  4: --
///  5: --
///  6: --
///  7: --
///  8: --
///  9: --
/// 10: |||||||||||||||||||| (100.00%)
/// 11: --
/// 12: --
/// 13: --
/// 14: --
/// 15: --
/// 16: --
///
/// Minimum Hamming distance: 10
/// Maximum Hamming distance: 10
///```
#[derive(Clone,Copy,Eq,PartialEq,Ord,ParitalOrd,Hash,Debug)]
#[repr(transparent)]
struct MyState(u16);

impl MyState {
  const MY_STATE0: MyState = MyState(0xcce9);
  const MY_STATE1: MyState = MyState(0x37d1);
  const MY_STATE2: MyState = MyState(0xf127);
}

</details>

### CODE REVIEW

**Review by msfschaffner — APPROVED** [review]
Thanks @lenary, this LGTM!

**Review by mcy — COMMENTED** [review]
A few questions:
- Is there a reason the info comment isn't doc comments? Seems like it would make more sense, esp. since I'm hoping to turn on `-D missing_docs` in our Rust code. E.g.
```
/// ```text
///  Encoding generated with...
/// ```
```
- Why is this a struct rather than a `#[repr(uN)]` enum? it should be trivial to generate code that goes from `uN` to enum value (you can go from enum to struct via an `as` cast). If the desire is to be able to represent states outside of those enumerated (for error-checking), I think the correct thing would be to use an enum and use `MaybeUninit<MyState>` for unverified values.
- Could you include `Debug, PartialOrd, Ord, Hash` in the derive? It's just good practice for things that are morally `enum class` from C++.

**Comment by lenary** [comment]
> * ~Why is this a struct rather than a `#[repr(uN)]` enum? it should be trivial to generate code that goes from `uN` to enum value (you can go from enum to struct via an `as` cast).~ If the desire is to be able to represent states outside of those enumerated (for error-checking), the correct thing would be to keep the enum and use `MaybeUninit<MyState>` for unverified values, which is consistent with other uses of MaybeUninit.

I don't think this is right at all. `MaybeUninit<T>` says "this starts off in an unknown state, and then later becomes initialized" - whereas we're using these types for "at no time can you assume a register/memory contains a known good value, or that value will remain a known good value", and so you always have to be able to pattern match on the value to check it's one of a few good values in many more bad values. Given you're not allowed to read a `MaybeUninit<T>` that's uninitalized, and if you could, you cannot check that it's one of the known good values (because you have no control over the representation anyway). This approach steps around that entirely.

**Comment by lenary** [comment]
@mcy Added the `#[derive(…)]`s you wanted, and the comment re-format.

**Comment by mcy** [comment]
> whereas we're using these types for "at no time can you assume a register/memory contains a known good value, or that value will remain a known good value"

It's not clear what this means. If at no time it can be assumed to be a good value, how can it be meaningfully used at all? Do we *actually* expect GPRs to be glitched? My understanding is that we did not.

Since this wasn't clear from my comment, I would expect a function like the following the be provided:
```rs
pub fn try_read(maybe: MaybeUninit<MyEnum>) -> Option<MyEnum> {
  match (maybe.as_ptr() as *const uN).read() {
    /* vald values */ => Some(maybe.read()),
    _ => None,
  }
}
```
The Rust aliasing rules permit the above so long as the pointer is well-aligned, which is guaranteed by the representation. Are we expecting values to get glitched after we read them into GPRs?

But in any case your solution works if all you want is pattern-matching. Make sure to mark it as `#[repr(transparent)]` so it can be safely sent over FFI boundaries and whatnot.

**Comment by lenary** [comment]
The point is I want you to pattern match on the value each time it is used, and it's not clear that the `try_read` is better than just an explicit pattern match.

**Comment by lenary** [comment]
Added `#[repr(transparent)]`

**Review by mcy — APPROVED** [review]
I don't *love* it, but this solution works. LGTM

File: hw/ip/prim/util/sparse-fsm-encode.py [hunk 1]
```diff
@@ -196,6 +196,10 @@ def main():
         pop_cnt = bin(c).count('1')
         if pop_cnt < args.n and pop_cnt > 0:
             for k in encodings:
+                # disalow candidates that are the complement of other states
```
Comment (imphil) [inline hw/ip/prim/util/sparse-fsm-encode.py:199]:
```suggestion
                # disallow candidates that are the complement of other states
```

Comment (lenary) [reply hw/ip/prim/util/sparse-fsm-encode.py:199]: 😀 ooops

**Comment by gkelly** [comment]
> The point is I want you to pattern match on the value each time it is used, and it's not clear that the `try_read` is better than just an explicit pattern match.

Right, this is an important distinction. There's a much larger window for attack if this value gets validated once and then implicitly trusted for its lifetime. GPRs are in scope for attack, but even if they weren't there's no guarantee this data doesn't spill to the stack at some point.

**Review by gkelly — APPROVED** [review]

**Pull request merged**